File: shell/macos/Sources/HeadroomKit/Models/Usage.swift
public struct Usage: Decodable, Sendable, Hashable {
    public let provider: String
    public let providerName: String
    public let usageHome: String
    public let today: UsageTotals
    public let yesterday: UsageTotals
    public let last30Days: UsageTotals
    public let daily: [DailyUsage]

    enum CodingKeys: String, CodingKey {
        case provider, today, yesterday, daily
        case providerName = "provider_name"
        case usageHome = "usage_home"
        case last30Days = "last_30_days"
    }
}

public struct TokenCounts: Decodable, Sendable, Hashable {
    public let input: UInt64
    public let cacheRead: UInt64
    public let cacheWrite: UInt64
    public let output: UInt64
    public let reasoning: UInt64
    public let total: UInt64

    enum CodingKeys: String, CodingKey {
        case input, output, reasoning, total
        case cacheRead = "cache_read"
        case cacheWrite = "cache_write"
    }
}

public struct UsageTotals: Decodable, Sendable, Hashable {
    public let tokens: TokenCounts
    public let costUSDMicros: Int64
    public let partial: Bool
    public let unpricedTokens: UInt64
    public let unpricedModels: [String]
    public let models: [ModelUsage]
    public let modelsOther: OtherModels?

    enum CodingKeys: String, CodingKey {
        case tokens, partial, models
        case costUSDMicros = "cost_usd_micros"
        case unpricedTokens = "unpriced_tokens"
        case unpricedModels = "unpriced_models"
        case modelsOther = "models_other"
    }
}

public struct DailyUsage: Decodable, Sendable, Hashable {
    public let date: String
    public let totalTokens: UInt64
    public let costUSDMicros: Int64
    public let partial: Bool

    enum CodingKeys: String, CodingKey {
        case date, partial
        case totalTokens = "total_tokens"
        case costUSDMicros = "cost_usd_micros"
    }
}

public struct ModelUsage: Decodable, Sendable, Hashable {
    public let model: String
    public let totalTokens: UInt64
    public let costUSDMicros: Int64
    public let partial: Bool
    public let costPerMTokUSDMicros: Int64?

    enum CodingKeys: String, CodingKey {
        case model, partial
        case totalTokens = "total_tokens"
        case costUSDMicros = "cost_usd_micros"
        case costPerMTokUSDMicros = "cost_per_mtok_usd_micros"
    }
}

public struct OtherModels: Decodable, Sendable, Hashable {
    public let count: UInt64
    public let totalTokens: UInt64
    public let costUSDMicros: Int64
    public let partial: Bool

    enum CodingKeys: String, CodingKey {
        case count, partial
        case totalTokens = "total_tokens"
        case costUSDMicros = "cost_usd_micros"
    }
}

public struct Spend: Decodable, Sendable, Hashable {
    public let today: PeriodSpend
    public let yesterday: PeriodSpend
    public let last7Days: PeriodSpend?
    public let last30Days: PeriodSpend

    enum CodingKeys: String, CodingKey {
        case today, yesterday
        case last7Days = "last_7_days"
        case last30Days = "last_30_days"
    }

    public func period(_ period: SpendPeriodPreference) -> PeriodSpend? {
        switch period {
        case .today: today
        case .yesterday: yesterday
        case .last7Days: last7Days
        case .last30Days: last30Days
        }
    }
}

public struct PeriodSpend: Decodable, Sendable, Hashable {
    public let costUSDMicros: Int64
    public let totalTokens: UInt64
    public let partial: Bool
    public let byProvider: [ProviderSpend]
    public let costPerMTokUSDMicros: Int64?
    public let projects: [ProjectSpend]?
    public let projectsOther: OtherProjects?

    enum CodingKeys: String, CodingKey {
        case partial, projects
        case costUSDMicros = "cost_usd_micros"
        case totalTokens = "total_tokens"
        case byProvider = "by_provider"
        case costPerMTokUSDMicros = "cost_per_mtok_usd_micros"
        case projectsOther = "projects_other"
    }
}

public struct ProviderSpend: Decodable, Sendable, Hashable {
    public let provider: String
    public let providerName: String
    public let costUSDMicros: Int64
    public let totalTokens: UInt64
    public let partial: Bool
    public let models: [ModelUsage]
    public let modelsOther: OtherModels?
    public let costPerMTokUSDMicros: Int64?

    enum CodingKeys: String, CodingKey {
        case provider, partial, models
        case providerName = "provider_name"
        case costUSDMicros = "cost_usd_micros"
        case totalTokens = "total_tokens"
        case modelsOther = "models_other"
        case costPerMTokUSDMicros = "cost_per_mtok_usd_micros"
    }
}

File: crates/headroom-tray/src/payload/usage.rs
use serde::Deserialize;

use super::SpendPeriod;
use super::projects::{OtherProjects, ProjectSpend};

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct Usage {
    pub provider: String,
    pub provider_name: String,
    pub usage_home: String,
    pub today: Totals,
    pub yesterday: Totals,
    pub last_30_days: Totals,
    pub daily: Vec<Daily>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct Totals {
    pub tokens: Tokens,
    pub cost_usd_micros: i64,
    pub partial: bool,
    pub models: Vec<ModelUsage>,
    pub models_other: Option<OtherModels>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct Tokens {
    #[serde(default)]
    pub input: u64,
    #[serde(default)]
    pub cache_read: u64,
    #[serde(default)]
    pub cache_write: u64,
    #[serde(default)]
    pub output: u64,
    #[serde(default)]
    pub reasoning: u64,
    pub total: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct Daily {
    pub date: jiff::civil::Date,
    pub total_tokens: u64,
    pub cost_usd_micros: i64,
    pub partial: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct ModelUsage {
    pub model: String,
    pub total_tokens: u64,
    pub cost_usd_micros: i64,
    pub partial: bool,
    #[serde(default)]
    pub cost_per_mtok_usd_micros: Option<i64>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct OtherModels {
    pub count: u64,
    pub total_tokens: u64,
    pub cost_usd_micros: i64,
    pub partial: bool,
    #[serde(default)]
    pub cost_per_mtok_usd_micros: Option<i64>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Deserialize)]
pub struct Spend {
    pub today: PeriodSpend,
    pub yesterday: PeriodSpend,
    #[serde(default)]
    pub last_7_days: Option<PeriodSpend>,
    pub last_30_days: PeriodSpend,
}

impl Spend {
    #[must_use]
    pub fn period(&self, period: SpendPeriod) -> Option<&PeriodSpend> {
        match period {
            SpendPeriod::Today => Some(&self.today),
            SpendPeriod::Yesterday => Some(&self.yesterday),
            SpendPeriod::Last7Days => self.last_7_days.as_ref(),
            SpendPeriod::Last30Days => Some(&self.last_30_days),
        }
    }

    #[must_use]
    pub fn has_period(&self, period: SpendPeriod) -> bool {
        self.period(period).is_some()
    }

    #[must_use]
    pub fn has_projects(&self) -> bool {
        self.today.projects.is_some()
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Deserialize)]
pub struct PeriodSpend {
    pub cost_usd_micros: i64,
    pub total_tokens: u64,
    pub partial: bool,
    pub by_provider: Vec<ProviderSpend>,
    #[serde(default)]
    pub cost_per_mtok_usd_micros: Option<i64>,
    #[serde(default)]
    pub projects: Option<Vec<ProjectSpend>>,
    #[serde(default)]
    pub projects_other: Option<OtherProjects>,
    #[serde(default)]
    pub models: Option<Vec<ProviderModelUsage>>,
    #[serde(default)]
    pub models_other: Option<OtherModels>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct ProviderModelUsage {
    pub provider: String,
    pub provider_name: String,
    #[serde(flatten)]
    pub usage: ModelUsage,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct ProviderSpend {
    pub provider: String,
    pub provider_name: String,
    pub cost_usd_micros: i64,
    pub total_tokens: u64,
    pub partial: bool,
    pub models: Vec<ModelUsage>,
    pub models_other: Option<OtherModels>,
    #[serde(default)]
    pub cost_per_mtok_usd_micros: Option<i64>,
}
